src/main/scala/firrtl/RenameMap.scala

```diff
@@ -178,124 +178,257 @@ final class RenameMap private () {
+              _.addHierarchy(t.moduleOpt.get, outerInst)
+            })
+        }
+      }
+    }
 
-        // If t is a ModuleTarget, try to rename parent target, then update t's parent
-        case t: ModuleTarget => getter(t.targetParent).map {
-          case CircuitTarget(c) => ModuleTarget(c, t.module)
+    def traverseRight(key: InstanceTarget): Seq[IsModule] = {
+      val findLeft = traverseLeft(key)
+      if (findLeft.nonEmpty) {
+        findLeft.get
+      } else {
+        key match {
+          case t: InstanceTarget if t.isLocal => Seq(key)
+          case t: InstanceTarget =>
+            val (Instance(i), OfModule(m)) = t.path.last
+            val parent = t.copy(path = t.path.dropRight(1), instance = i, ofModule = m)
+            traverseRight(parent).map(_.instOf(t.instance, t.ofModule))
         }
+      }
+    }
 
-        /** If t is an InstanceTarget (has a path) but has no references:
-          * 1) Check whether the instance has been renamed (asReference)
-          * 2) Check whether the ofModule of the instance has been renamed (only 1:1 renaming is ok)
-          */
-        case t: InstanceTarget =>
-          getter(t.asReference).map {
-            case t2:InstanceTarget => t2
-            case t2@ReferenceTarget(c, m, p, r, Nil) =>
-              val t3 = InstanceTarget(c, m, p, r, t.ofModule)
-              val ofModuleTarget = t3.ofModuleTarget
-              getter(ofModuleTarget) match {
-                case Seq(ModuleTarget(newCircuit, newOf)) if newCircuit == t3.circuit => t3.copy(ofModule = newOf)
-                case other =>
-                  errors += s"Illegal rename: ofModule of $t is renamed to $other - must rename $t directly."
-                  t
-              }
-            case other =>
-              errors += s"Illegal rename: $t has new instance reference $other"
-              t
+    traverseRight(key)
+  }
+
+  private def circuitGet(errors: mutable.ArrayBuffer[String])(key: CircuitTarget): Seq[CircuitTarget] = {
+    underlying.get(key).map(_.flatMap {
+      case c: CircuitTarget => Some(c)
+      case other =>
+        errors += s"Illegal rename: $key cannot be renamed to non-circuit target: $other"
+        None
+    }).getOrElse(Seq(key))
+  }
+
+  private def moduleGet(errors: mutable.ArrayBuffer[String])(key: ModuleTarget): Seq[IsModule] = {
+    underlying.get(key).map(_.flatMap {
+      case mod: IsModule => Some(mod)
+      case other =>
+        errors += s"Illegal rename: $key cannot be renamed to non-module target: $other"
+        None
+    }).getOrElse(Seq(key))
+  }
+
+  /** Recursively renames a target so the returned targets are complete renamed
+    * @param errors Used to record illegal renames
+    * @param key Target to rename
+    * @return Renamed targets
+    */
+  private def recursiveGet(errors: mutable.ArrayBuffer[String])(key: CompleteTarget): Seq[CompleteTarget] = {
+    if(getCache.contains(key)) {
+      getCache(key)
```

Comment:
Can we cache the other recursive calls like traverse right and traverse left? I think this could be important if lots of signals of a deep instance get renamed (or something like that). I'm open to not caching those functions as well, if you feel it has limited use.